vllm/core/scheduler.py

```diff
@@ -310,6 +321,98 @@ def scheduled_seq_group_builder():
     # return ScheduledSequenceGroup(seq_group=None, token_chunk_size=0)
 
 
+@dataclass
+class PartialPrefillMetadata:
+    """Holds information about the partial prefills that are currently running
+    during a single iteration of the Scheduler.
+    When chunked prefill is enabled, we allow a certain number of seqs to be
+    partially prefilled during each iteration. Having multiple partial prefills
+    in flight allows us to minimize TTFT and avoid decode starvation in cases
+    where a single sequence group with a very large prompt blocks the queue for
+    too many iterations.
+    The number of long prefill requests is limited so that smaller
+    requests may jump the queue in front of them and get to the decode
+    phase faster.
+    """
+
+    # A minimum bound on the total number of prefills running during this
+    # scheduling step
+    partial_prefills: int
```

Comment:
just a preliminary idea: do you think it would be possible to calculate the number of partial running prefills as we were doing the scheduling from the previous iteration and just track as part of scheduler's internal state?

Reply:
> just a preliminary idea: do you think it would be possible to calculate the number of partial running prefills as we were doing the scheduling from the previous iteration and just track as part of scheduler's internal state?

Hah, that's what we were doing before but Cody suggested calculating and passing this metadata down to the scheduling methods where it's needed. So the goal here is to calculate just enough info at the beginning of each scheduler iteration to properly budget the partial prefills, hence the early breaks from looping through the queues.

> oh, we are also early breaking the loop, so this variable is kind of weird? it's then num_schedulable_prefills where shcedulable could be running and waiting?

Yeah that's an accurate name, maybe we can use that